### DISCUSSION TITLE
Inconsistent JSON performance for Rust frameworks.

### DISCUSSION THREAD

**Discussion by CordlessCoder** [TechEmpower/FrameworkBenchmarks#8634] [General]
Some Rust frameworks i.e. ntex, may-minihttp are using the extremely performant `yarte` for JSON serialization, while others i.e. axum, actix are using the far slower serde.

I believe we should make sure all Rust frameworks use yarte for serialization to make this a fair comparison.

**Comment by fakeshadow** [comment]
The frameworks you mentioned all depend on `yarte` already so if you find it useful for json you can just make PR for them. 

I would not call it inconsistency and I strongly against making `yarte` mandatory. Personally I use `serde_json` because it's widely used in real world and performs good enough. It's also lightweight on dependency result in fast build time which I value more than some additional score points.

**Reply by CordlessCoder** [reply]
Why not call it inconsistency when different frameworks are using entirely different libraries for serializing JSON, with vastly different performance? Why not reduce the number of variables by only using `yarte` for the JSON endpoint? I see no argument against it, unless you want to claim using `serde_json` is more realistic in which case every framework should be switched from yarte to serde_json.

**Reply by fakeshadow** [reply]
Because there are multiple approaches to the same problem with different outcomes and trade off. I already explained why I use serde_json in the bench instead of yarte and if you look on the other side of it another person may find he wants the opposite.

It comes down to not forcing a single library on everyone and give people freedom of choices.

**Reply by fakeshadow** [reply]
To reinforce my point here is a reference of binary build time comparing `actix-web` and `xitca-web` of the offical round 22:
```
actix-http:     Finished release [optimized] target(s) in 4m 52s
xitca-web:     Finished release [optimized] target(s) in 43.22s
```
To be clear I'm not saying longer build time is bad nor it's a fair comparison(this is a completely cold build and actix in general build a lot of useless dependency that certain binary does not need). And it's definitely not solely `yarte` being the cause of longer build time. But it can clearly show you how different approaches have different outcome with unique trade off and using `serde_json` 100% contribute to my approach of faster build time.

**Reply by CordlessCoder** [reply]
"Multiple approaches to the same problem" - but we're not benchmarking those approaches, we're benchmarking the web frameworks! The serialization framework you use isn't part of the web framework, so all you're doing is adding one more variable to make the results less directly comparable. I see absolutely 0 reason for the inconsistency, and yes, I will make a PR

**Reply by fakeshadow** [reply]
What you said is wrong on two fronts:
1. Framework benchmark is not only about score. There are many aspects to it and people have drastically different takes on what they saw from tfb benchmark based on their opinion. Like I said multiple times you can score how much and I don't give a f**k if it contradict my value and vice versa. So the best thing to do is to give people freedom and respect the difference.
2. serializer is part of a high level web framework. Just take a look at the frameworks you listed: [actix-web](https://github.com/actix/actix-web/blob/1114a51b222e017c530b3d09a1cf595729205e15/actix-web/Cargo.toml#L99), [axum](https://github.com/tokio-rs/axum/blob/2b486ea174b263e4335490794637ea50eb04765c/axum/Cargo.toml#L59),  and even [ntex](https://github.com/ntex-rs/ntex/blob/a4f9802d6d23db5f5972d20e06398997ad2a7f7a/ntex/Cargo.toml#L87). Of course you can bring in your own json (de)serializer but it does not change the fact that their offer a default. (`ntex` being an example of this where one can end up look at the `yarte` score while using `serde_json` without knowing)

**Reply by fakeshadow** [reply]
Hey look at how fast xxx framework is with blazing fast json serializer. I'm gonna try it:
```rust
async fn json(_: Json<Foo>) -> &'static str{
    "no you are actually using serde_json"
}
```
This could be a bad joke and I mean no offense to the related maintainers of xxx as it's a common trick lots of people use to boost their scores in the benchmark regardless languages. But personally I find it misleading at least for framework users.
